annotation-processor/src/main/java/org/hibernate/validator/ap/classchecks/MethodOverrideCheck.java

```diff
@@ -0,0 +1,274 @@
+/*
+ * Hibernate Validator, declare and validate application constraints
+ *
+ * License: Apache License, Version 2.0
+ * See the license.txt file in the root directory or <http://www.apache.org/licenses/LICENSE-2.0>.
+ */
+package org.hibernate.validator.ap.classchecks;
+
+
+import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
+import java.util.Set;
+import javax.lang.model.element.AnnotationMirror;
+import javax.lang.model.element.Element;
+import javax.lang.model.element.ElementKind;
+import javax.lang.model.element.ExecutableElement;
+import javax.lang.model.element.TypeElement;
+import javax.lang.model.type.TypeMirror;
+import javax.lang.model.util.Elements;
+import javax.lang.model.util.Types;
+
+import org.hibernate.validator.ap.checks.ConstraintCheckError;
+import org.hibernate.validator.ap.util.CollectionHelper;
+
+/**
+ * Abstract base class for {@link ClassCheck} implementations of overridden method checks.
+ *
+ * @author Marko Bekhta
+ */
+public abstract class MethodOverrideCheck extends AbstractClassCheck {
+
+	private static final String JAVA_LANG_OBJECT = "java.lang.Object";
+
+	private static final String JAVAX_VALIDATION_CONSTRAINT = "javax.validation.Constraint";
+
+	private final Elements elementUtils;
+
+	private final Types typeUtils;
+
+	public MethodOverrideCheck(Elements elementUtils, Types typeUtils) {
+		this.elementUtils = elementUtils;
+		this.typeUtils = typeUtils;
+	}
+
+	@Override
+	public Collection<ConstraintCheckError> checkMethod(ExecutableElement currentMethod) {
+		if ( !needToPerformAnyChecks( currentMethod ) ) {
+			return Collections.emptySet();
+		}
+		TypeElement currentTypeElement = getEnclosingTypeElement( currentMethod );
+
+		// find if there's a method that was overridden by a current one.
+		List<ExecutableElement> overridden = findAllOverriddenElements( currentTypeElement, currentMethod );
+		if ( overridden.isEmpty() ) {
+			return Collections.emptySet();
+		}
+
+		// if there's more than one overridden method we need to make sure that all of them match
+		Set<ConstraintCheckError> errors = CollectionHelper.newHashSet();
+		for ( ExecutableElement firstExecutable : overridden ) {
```

Comment:
(Sorry for the delay on this one, it took me some time to find some spare cycles to fully dedicate to this review)

I might be wrong but I don't think this is right. You're comparing each method with the others while the order is important and you should only compare an overriding method to the ones it overrides not the other way around.

I think what would be in order is to build several separate inheritance trees for each path and then runs the check on each inheritance tree in the right order. You might have several inheritance trees because of the interfaces. Doing it like that will also make the detection of parallel declarations robust.

I think you could trigger the issue by creating an inheritance with several super classes.

Did I miss something? Does what I say make sense to you?